sphinx/texinputs/sphinxlatexnumfig.sty

```diff
@@ -101,20 +101,20 @@
     \ifspx@opt@mathnumfig
       \@addtoreset{equation}{subparagraph}%
     \fi%
-    \g@addto@macro\spx@preAthefigure{\ifnum\c@subsubparagraph>\z@\arabic{subsubparagraph}.}%
+    \g@addto@macro\spx@preAthefigure{\ifnum\c@subsubparagraph>\z@.\arabic{subsubparagraph}}%
     \g@addto@macro\spx@preBthefigure{\fi}%
 \fi
 \expandafter\g@addto@macro
 \expandafter\spx@preAthefigure\expandafter{\spx@preBthefigure}%
 \let\thefigure\spx@preAthefigure
 \let\thetable\spx@preAthefigure
 \let\theliteralblock\spx@preAthefigure
-\g@addto@macro\thefigure{\arabic{figure}}%
-\g@addto@macro\thetable{\arabic{table}}%
-\g@addto@macro\theliteralblock{\arabic{literalblock}}%
+\g@addto@macro\thefigure{.\arabic{figure}}%
+\g@addto@macro\thetable{.\arabic{table}}%
+\g@addto@macro\theliteralblock{.\arabic{literalblock}}%
   \ifspx@opt@mathnumfig
     \let\theequation\spx@preAthefigure
-    \g@addto@macro\theequation{\arabic{equation}}%
+    \g@addto@macro\theequation{\spx@opt@mathnumsep\arabic{equation}}%
```

Comment:
I need to go check the full file to remember its logic.  It may be possible to keep alls dots in place, replacing them by some `\spx@dot@checkifnextisequation` and inject `\spx@opt@mathnumsep` only then, but I need to check more.  I am bit worried about putting the dots upfront in case there is some reason not obvious to see about how they are positioned now.

Reply:
As I read the file, the flow of logic works either way. I implemented it this way to avoid adding even more logic (e.g., `\spx@dot@checkifnextisequation`).

Reply:
It does look a priori more logic to append a sublevel with dot separator first.  But there are good reasons why the file was structured as it was with the dot on the right, inside the nested TeX conditionals.  It is only via re-establishing it that way, with a special dot macro that I could solve all cases of use, cf [this comment](https://github.com/sphinx-doc/sphinx/pull/12652#pullrequestreview-2192541107). If we add it first we have to also add a layer to remove the first one if `latex_toplevel_sectioning='part'` but that's not only problem.

On the latex side, a notable thing is that the construct must be compatibly with expansion context.  Hyperref submit it to a `\protected@edef`. And even without hyperref, it is expanded when written to `.aux` file.  So any thing using `\def` or `\edef` is forbidden.

Here is the unrelated to `numsep` prior goal of the construct in this file.  A user can put an equation at any level of depth: it can appear immediately under what will become a `\chapter` or a `\section`... or a `\subpararagraph`.  etc... If you define `\theequation` to match numfigreset, for example equations resets their numbering at each `\subsection` the simple way for typical LaTeX user will be do something like `\renewcommand\theequation{\thesubsection.\arabic{equation}` (or whatever separator you want).  The problem is that if you equation is near start of chapter 1, not contained in any section you now get `1.0.0.<eqnumber>`.  But the zeros refer to non existent entity.  What the construct in this file achieve is that you will get rather `1.<eqnumber>`.  

Parenthetical paragraph: It is indeed true that then with using only `dot` as sep, we have possibly confusing reference, except that in LaTeX+amsmath used by Sphinx we use `\eqref` which adds parentheses absent of other reference. And it is surely possibly to configure/define `\eqref` to say thing such as "equation <>", so this is how things could be done.  But the idea of your PR makes it easier indeed. And I see it is motivated from HTML where the same confusion could arise with labels.

Now going back to paragraph before, one could think, ok use the value of numfigreset and define for example as `\renewcommand\theequation{\thesubsection.\arabic{equation}` and get rid of the zeroes afterwards. The "removal of zeroes" would have needed back in the ages TeX wizardry about the same level of the type I had to introduce to let your PR work.  As I said we can't do `\edef` or `\protected@edef` ourselves.  The `\expanded` primitive was only introduced recently (2019-2020) to LaTeX.  But if the file had been structured via such a strategy, adding your `numsep` would have been easier.

Things being as they are, I implemented the simplest way to make it work in general.